pkg/maglev/maglev.go

```diff
@@ -4,80 +4,141 @@
 package maglev
 
 import (
+	"cmp"
 	"context"
 	"encoding/base64"
 	"fmt"
+	"iter"
 	"runtime"
 	"slices"
+	"strconv"
+	"strings"
 
+	"github.com/cilium/hive/cell"
 	"github.com/cilium/workerpool"
-	"github.com/mackerelio/go-osstat/memory"
+	"github.com/spf13/pflag"
 
 	"github.com/cilium/cilium/pkg/loadbalancer"
+	"github.com/cilium/cilium/pkg/lock"
 	"github.com/cilium/cilium/pkg/murmur3"
 )
 
+var Cell = cell.Module(
+	"maglev",
+	"Maglev table computations",
+
+	cell.Config(DefaultConfig),
+	cell.Provide(New),
+)
+
 const (
 	DefaultTableSize = 16381
 
 	// seed=$(head -c12 /dev/urandom | base64 -w0)
 	DefaultHashSeed = "JLfvgnHc2kaSUFaI"
+
+	MaglevTableSizeName = "bpf-lb-maglev-table-size"
+	MaglevHashSeedName  = "bpf-lb-maglev-hash-seed"
 )
 
-var (
-	seedMurmur uint32
+type Config struct {
+	// Maglev backend table size (M) per service. Must be prime number.
+	MaglevTableSize uint `mapstructure:"bpf-lb-maglev-table-size"`
+
+	// MaglevHashSeed contains the cluster-wide seed for the hash(es).
+	MaglevHashSeed string `mapstructure:"bpf-lb-maglev-hash-seed"`
+}
+
+func (def Config) Flags(flags *pflag.FlagSet) {
+	flags.Uint(MaglevTableSizeName, def.MaglevTableSize, "Maglev per service backend table size (parameter M)")
+	flags.String(MaglevHashSeedName, def.MaglevHashSeed, "Maglev cluster-wide hash seed (base64 encoded)")
+}
 
+var DefaultConfig = Config{
+	MaglevTableSize: DefaultTableSize,
+	MaglevHashSeed:  DefaultHashSeed,
+}
+
+type Maglev struct {
+	Config
 	SeedJhash0 uint32
 	SeedJhash1 uint32
+	seedMurmur uint32
 
-	// permutation is the slice containing the Maglev permutation calculations.
-	permutation []uint64
-)
+	// mu protects the fields below
+	mu lock.Mutex
+
+	wp *workerpool.WorkerPool
+
+	// backendInfosBuffer is a reusable buffer for holding the backend infos.
+	backendInfosBuffer []BackendInfo
+
+	// permutations is (re)used during each GetLookupTable call to compute the table.
+	permutations []uint64
+}
+
+// New constructs a new Maglev computation object.
+func New(cfg Config) (*Maglev, error) {
+	// "Let N be the size of a VIP's backend pool." [...] "In practice, we choose M to be
+	// larger than 100 x N to ensure at most a 1% difference in hash space assigned to
+	// backends." (from Maglev paper, page 6)
+	supportedPrimes := []uint{251, 509, 1021, 2039, 4093, 8191, 16381, 32749, 65521, 131071}
+	found := false
+	for _, prime := range supportedPrimes {
```

Comment:
nit: `slices.Index(..)` might be a bit cleaner

Reply:
`slices.Contains` is even cleaner.